lib/oauth/oauth.ex

```diff
@@ -24,4 +61,20 @@ defmodule OAuth do
       _ -> receive_code()
     end
   end
+
+  defp get_from_cache(get_fn) do
+    Agent.get(agent(), get_fn)
+  end
+
+  defp store_in_cache(update_fn) do
+    Agent.update(agent(), update_fn)
+  end
+
+  # Use a local agent as a cache system
+  defp agent() do
+    case Agent.start_link(fn -> %{} end, name: __MODULE__) do
```

Comment:
You typically should avoid starting "static" processes on the fly, this is the role the supervision tree 😉
If you use a named process, you don't need to even retrieve the `pid` each time, you can just use the process name [directly](https://hexdocs.pm/elixir/Agent.html#module-examples): `Agent.update(__MODULE__, update_fn)`

Reply:
Actually this cache might not work as you expect:
- calling start_link within the process handling a web request
- the agent is created and the two processes get linked: agent <=> request
- the request finished, taking down the request and the agent together
- the next request comes in: starts a new empty agent :p

Reply:
> You typically should avoid starting "static" processes on the fly, this is the role the supervision tree wink If you use a named process, you don't need to even retrieve the `pid` each time, you can just use the process name [directly](https://hexdocs.pm/elixir/Agent.html#module-examples):
> 
> ```elixir
> Agent.update(__MODULE__, update_fn)
> ```

But I would still need to potentially create the agent before updating the state no?

Anyway, how would you create a on persistent cache system with a lifetime == my app process (or even shorter) ?

Reply:
This is where the supervision tree comes in.
You can check the [plug doc](https://hexdocs.pm/plug/readme.html#supervised-handlers) to see how to add your plug server to your application tree, and then add your agent to the children.
More links:
- https://hexdocs.pm/elixir/Agent.html#module-how-to-supervise
- https://elixir-lang.org/getting-started/mix-otp/supervisor-and-application.html

Reply:
OK. Updated/Gave it a try.